app/models/lesson_plan.rb

```diff
@@ -75,9 +71,28 @@ def files
     end.uniq
   end
 
-  private
+  def to_param
+    persisted? ? id : "current"
+  end
+
+  def key!
+    LessonPlanLink.find_or_create_by(lesson_ids: lessons.pluck(:id)).key
+  end
+
+  def self.find_or_create_by_key(key)
+    link = LessonPlanLink.find_by!(key: key)
+    return link.lesson_plan if link.lesson_plan
 
-  def set_key
-    self.key = EffDiceware.generate(4).gsub(" ", "-")
+    LessonPlan.create.tap do |plan|
+      link.update_attribute(:lesson_plan, plan)
+      plan.lesson_ids = link.lesson_ids
+    end
+  end
+
+  def force_immutable_if_linked
+    if self.lesson_plan_link && self.persisted? && self.changed?
+      errors.add(:base, :immutable)
```

Comment:
This is clever. There's also [readonly!](https://apidock.com/rails/ActiveRecord/Core/readonly%21) which would raise an exception instead.

Reply:
I like that.

Reply:
Looks like `readonly!` doesn't protect against updating associations stored on another table. `plan.save` raises the error, but `plan.update_attributes(:lessons, [])` does not.